### DISCUSSION TITLE
Edge Functions: how to authorize the user AND bypass RLS?

### DISCUSSION THREAD

**Discussion by MartinMalinda** [supabase/supabase#15631] [Questions] [answered]
I'm using an Edge Function primarily to bypass RLS. There's a hidden table about a user that is intentionally not exposed. Perhaps I could split it and partially make it readable to the user but there's hidden "meta" parts that should only be changed by the system.

So in my edge function I need to identify the user. I can use the client for this purpose:

```ts
  const supabaseClient = createClient(
    Deno.env.get('SUPABASE_URL') ?? '',
    Deno.env.get('SUPABASE_ANON_KEY') ?? '',
    // This way your row-level-security (RLS) policies are applied.
    { global: { headers: { Authorization: req.headers.get('Authorization')! } } }
  )
 const { data: { user } } = await supabaseClient.auth.getUser();
```

This is nice, now I have `user.id` and I know it's really the user and not an impersonator. (Thanks to JWT I assume).

But from this point onward I'd like to bypass RLS. Do whatever needed basically access and change data I couldn't from the supabasejs client directly.

Do I need to just create a new separate client that uses the service key? Or is there a better way? Or should I pool the Postgres DB directly?

Thanks!

**Comment by GaryAustin1** [comment, marked as answer]
Just create another client with service role key.
You could also validate the jwt/access token directly with a jwt library and the Supabase jwt secret key and avoid the getUser call and other client if you really wanted to.

**Reply by MartinMalinda** [reply]
OK great, I wanted to double check this is a good practice. I can probably abstract the authorization to some oneliner function so it seems fine, thank you

**Comment by ivaancruzz** [comment]
> I'm using an Edge Function primarily to bypass RLS. There's a hidden table about a user that is intentionally not exposed. Perhaps I could split it and partially make it readable to the user but there's hidden "meta" parts that should only be changed by the system.
> 
> So in my edge function I need to identify the user. I can use the client for this purpose:
> 
> ```ts
>   const supabaseClient = createClient(
>     Deno.env.get('SUPABASE_URL') ?? '',
>     Deno.env.get('SUPABASE_ANON_KEY') ?? '',
>     // This way your row-level-security (RLS) policies are applied.
>     { global: { headers: { Authorization: req.headers.get('Authorization')! } } }
>   )
>  const { data: { user } } = await supabaseClient.auth.getUser();
> ```
> 
> This is nice, now I have `user.id` and I know it's really the user and not an impersonator. (Thanks to JWT I assume).
> 
> But from this point onward I'd like to bypass RLS. Do whatever needed basically access and change data I couldn't from the supabasejs client directly.
> 
> Do I need to just create a new separate client that uses the service key? Or is there a better way? Or should I pool the Postgres DB directly?
> 
> Thanks!

Your example helped me enable RLS on edge functions. Thanks! 🚀

**Reply by GaryAustin1** [reply]
You can create a new client with with service_role key and use the user id from getUser.   Just don't have load the auth header with a user jwt or it will then not be service_role.
